File: src/operand.cpp
#include "operand.hpp"
#include "instruction.hpp"
#include "function.hpp"
#include "executable.hpp"

Operand::Operand(cs_x86_op op): m_op(op), m_instruction(NULL), m_label("")
{}

void Operand::setInstruction(Instruction* instruction)
{
    if(instruction != NULL)
        m_instruction = instruction;
}
std::string Operand::toString()
{
    if(m_label != "")
        return m_label;
    
    switch(m_op.type)
    {
        case 2:
            return "$"+m_op.imm;
            break;
        case 1:
            return "%"+cs_reg_name(m_op.reg);
            break;
        case 4:
            return m_op.fp;
        case 3:
            std::string res;
            if(m_op.mem.disp != 0)
            {
                if(cs_reg_name(m_op.base) == "rip")
                {
                    int address = m_instruction->address + m_instruction->size + m_op.mem.disp;
                    bool isFunction = false;
                    for(int i(0);i < m_instruction->function()->executable()->functions.size;i++)
                        if(m_instruction->function()->executable()->functions[i].address == address)
                        {
                            isFunction = true;
                            res = m_instruction->function()->executable()->functions[i].name;
                            printf("The address %i is actualy the function %s", address, res);
                            if(m_instruction->function()->section.name == ".rodata" | m_instruction->function()->section.name == ".data")
                                m_instrucion->function()->executable()->addFunctionData(m_instruction->function()->executable()->functions[i]);
                        }

                    if(!isFunction)
                    {
                        res = ".LC"+m_instruction->function()->executable()->functions_data.size;
                        m_instruction->function()->executable()->addFunctionData(Function(address, 0, res));
                    }
                }
                else
                {
                    res = m_op.mem.disp;
                }
            }

            if(m_op.mem.base != 0 & m_op.mem.index != 0 & m_op.mem.scale != 1)
            {
                res += "(";
                if(m_op.mem.base != 0)
                    res += "%" + cs_reg_name(m_op.mem.base);
                if(m_op.mem.index != 0)
                    res += ", " + cs_reg_name(m_op.mem.index);
                if(m_op.meme.scale != 1)
                    res += ", " + m_op.mem.scale;
                res += ")"
            }
            break;
        default:
            printf("operand type Unknow.");
            break;
    }
}

File: IOT/core/operand.py
from IOT.core.function import *

class Operand():
    def __init__(self, op):
        self.op = op
        self.label = ""
    def __str__(self):
        if self.label != "":
            return self.label
        elif self.op.type == 2:
            return "$"+self.op.value.imm.__str__()
        elif self.op.type == 1:
            return "%"+self.instruction.insn.reg_name(self.op.value.reg)
        elif self.op.type == 4:
            return self.op.value.fp.__str__()
        elif self.op.type == 3:
            str = ""
            if self.op.value.mem.disp != 0:
                str = self.op.value.mem.disp.__str__()
#                if self.instruction.insn.reg_name(self.op.value.mem.base) == "rip":
                addr = self.instruction.address + self.instruction.size + self.op.value.mem.disp
                isF=False
                for function in self.instruction.function.executable.functions:
                    if function.address == addr:
                        isF=True
                        str = function.name
                        print("The address of this memory", addr, "at ", function.name)
                        if function.section.name == ".rodata" or function.section.name == ".data":
                            function.executable.functions_data.append(function)
                if not isF:
                    name = ".LC"+(len(self.instruction.function.executable.functions_data)+1).__str__()
                    self.instruction.function.executable.functions_data.append(Function(addr, 0, name,self.instruction.function.executable))
                    str = name
            str = str + "("
            if self.op.value.mem.base != 0:
                str = str + "%" + self.instruction.insn.reg_name(self.op.value.mem.base)
            if self.op.value.mem.index != 0:
                str = str + ", %" + self.instruction.insn.reg_name(self.op.value.mem.index)
            if self.op.value.mem.scale != 1:
                str = str + ", " + self.op.value.mem.scale.__str__()
            str = str + ")"
            return str
        if self.op.type == 0:
            return "Error type"
        else:
            print(self.instruction.op_str())